awscrt/http.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#  http://aws.amazon.com/apache2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import _aws_crt_python
+from concurrent.futures import Future
+from enum import IntEnum
+from awscrt.io import ClientBootstrap, TlsConnectionOptions, SocketOptions
+
+
+class HttpClientConnection(object):
+    """
+    Represents an Http connection to a remote endpoint. Everything in this class is non-blocking.
+    """
+    __slots__ = ('_bootstrap', '_tls_connection_options', '_on_connection_shutdown', '_native_handle')
+
+    # don't call me, I'm private
+    def __init__(self, bootstrap, on_connection_shutdown, tls_connection_options):
+        assert isinstance(bootstrap, ClientBootstrap)
+        assert tls_connection_options is None or isinstance(tls_connection_options, TlsConnectionOptions)
+        assert on_connection_shutdown is not None
+
+        for slot in self.__slots__:
+            setattr(self, slot, None)
+
+        self._bootstrap = bootstrap
+        self._tls_connection_options = tls_connection_options
+        self._on_connection_shutdown = on_connection_shutdown
+        self._native_handle = None
+
+    @staticmethod
+    def new_connection(bootstrap, host_name, port, socket_options, on_connection_shutdown, tls_connection_options):
+        """
+        Initiates a new connection to host_name and port using socket_options and tls_connection_options if supplied.
+        if tls_connection_options is None, then the connection will be attempted over plain-text.
+        on_connection_shutdown will be invoked if the connection shuts-down while you still hold a reference to it.
+        on_connection_shutdown takes a single argument of int type, to specify the shutdown reason.
+
+        returns a future where the result is a new instance to HttpClientConnection, once the connection has completed
+        and is ready for use.
+        """
+        assert tls_connection_options is None or isinstance(tls_connection_options, TlsConnectionOptions)
+        assert host_name is not None
+        assert port is not None
+        assert socket_options is not None and isinstance(socket_options, SocketOptions)
+        assert on_connection_shutdown is not None
```

Comment:
Seems like we should allow `on_connection_shutdown` to be None
Which means it should also get a default `on_connection_shutdown=None` in the fn declaration